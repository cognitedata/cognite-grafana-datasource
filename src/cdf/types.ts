export type Datapoints = Items<Datapoint>;
export type Events = Items<Event>;

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export interface TimeSeriesDatapoint {
  timestamp: number;
  value: string;
}

export interface TimeSeriesAggregateDatapoint {
  timestamp: number;
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface CogniteEvent {
  id: number;
  lastUpdatedTime: string;
  createdTime: string;
  externalId?: string;
  startTime?: number;
  endTime?: number;
  dataSetId?: number;
  description?: string;
  type?: string;
  subtype?: string;
  assetIds?: number[];
  source?: string;
  sourceId?: string;
  metadata?: { [s: string]: string };
}

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  activeAtTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export type EventsFilterTimeParams =
  | Pick<EventsFilterRequestParams, 'activeAtTime'>
  | Pick<EventsFilterRequestParams, 'startTime' | 'endTime'>;

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter?: Filter;
  advancedFilter?: any;
}

export interface Resource {
  id: number;
  externalId?: string;
  name?: string;
  description?: string;
}

export interface TimeSeriesResponseItem extends Resource {
  isString?: boolean;
  metadata?: Record<string, string>;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface Timestamp {
  timestamp: number;
}

export type Items<T = any> = {
  items: T[];
};

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export interface Range<T> {
  min?: T;
  max?: T;
}

export interface CogniteLabelsResponse {
  externalId: string;
  name: string;
  description: string;
  dataSetId: number;
  createdTime: number;
}
interface CogniteRelationshipAsset {
  createdTime?: number;
  lastUpdatedTime?: number;
  rootId?: number;
  aggregates?: {
    childCount: number;
    depth: number;
    path?: [
      {
        id: string | number;
      }
    ];
  };
  parentId?: number;
  parentExternalId?: string;
  externalId: string;
  name?: string;
  description?: string;
  dataSetId: number;
  metadata?: { [s: string]: string };
  source?: string;
  labels?: CogniteLabelsResponse[];
  id: number;
}
export interface CogniteRelationshipResponse {
  externalId: string;
  sourceExternalId: string;
  sourceType?: string;
  targetExternalId: string;
  targetType?: string;
  startTime?: number;
  endTime?: number;
  confidence?: number;
  dataSetId?: number;
  labels?: CogniteLabelsResponse[];
  createdTime?: number;
  lastUpdatedTime?: number;
  source?: CogniteRelationshipAsset;
  target?: CogniteRelationshipAsset;
}
export interface RelationshipsFilter {
  dataSetIds?: {
    id: number;
    value?: string;
  }[];
  labels?: {
    containsAny: {
      externalId: string;
      value?: string;
    }[];
  };
  isActiveAtTime?: boolean;
  activeAtTime?: {
    max: number;
    min: number;
  };
  sourceExternalIds?: string[];
  targetTypes?: string[];
}
