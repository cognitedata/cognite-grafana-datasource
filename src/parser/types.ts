import { STSClientFilter } from './ts';

export const FilterTypes = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export type ParsedFilter = STSClientFilter;

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
