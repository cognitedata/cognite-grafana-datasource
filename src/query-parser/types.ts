export const FilterType = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';
export interface QueryFilter {
  filter: FilterType;
  value: string;
}

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export interface ParsedFilter extends QueryFilter {
  path: string;
}

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
