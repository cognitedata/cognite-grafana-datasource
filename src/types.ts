import {
  DataQueryOptions,
  DataQuery,
  TimeSeries,
  TimeRange as GrafanaTimeRange,
  DataSourceSettings,
} from '@grafana/ui';

export function isError(maybeError: DataQueryError | any): maybeError is DataQueryError {
  return (<DataQueryError>maybeError).error !== undefined;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 * @param name â€“ event name
 */
export const eventFactory = <T = undefined>(name: string): AppEvent<T> => {
  return { name };
};

export type QueryResponse = DataResponse<TimeSeries[]>;

export interface MetricDescription {
  readonly text: string;
  readonly value: number | string;
}

export type MetricFindQueryResponse = MetricDescription[];

export enum Tab {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Custom = 'Custom',
}

export enum ParseType {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Event = 'Event',
}

export interface TimeSeriesResponseItem {
  id: number;
  externalId?: string;
  name?: string;
  isString?: boolean;
  metadata?: object;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  description?: string;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export interface AssetQuery {
  target: string;
  includeSubtrees: boolean;
  old?: AssetQuery;
  timeseries?: TimeSeriesResponseItem[];
  func?: string;
  templatedTarget?: string;
}

export interface InputQueryTarget extends DataQuery {
  target: number | '';
  aggregation: string;
  granularity: string;
  error: string;
  label: string;
  tab: Tab;
  assetQuery: AssetQuery;
  expr: string;
  warning: string;
}

export interface QueryTarget extends InputQueryTarget {
  target: number;
}

export type QueryFormat = 'json';

export type QueryOptions = DataQueryOptions<InputQueryTarget>;

export type Tuple<T> = [T, T];

export interface Range<T> {
  min?: T;
  max?: T;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 */
export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export interface DataSourceRequestOptions {
  url: string;
  method: HttpMethod;
  retry?: number;
  requestId?: string;
  headers?: { [s: string]: string };
  silent?: boolean;
  data?: object;
}

export interface Timestamp {
  timestamp: number;
}

export type MetaResponses = Responses<ResponseMetadata, DataQueryRequestResponse>;

export type SuccessResponse<Metadata, Response> = { metadata: Metadata; result: Response };
export type FailResponse<Metadata> = { metadata: Metadata; error: any };
export type Responses<Metadata, Response> = {
  failed: FailResponse<Metadata>[];
  succeded: SuccessResponse<Metadata, Response>[];
};

export interface TimeSeriesDatapoint extends Timestamp {
  value: string;
}

export interface TimeSeriesAggregateDatapoint extends Timestamp {
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export type Datapoints = Items<Datapoint>;

export interface DataQueryRequestResponse extends DataResponse<Datapoints> {
  config: {
    data: {
      aggregates: string;
      limit: number;
    };
  };
}

export interface RequestParams<DataType = any> {
  path: string;
  data: DataType;
  method: HttpMethod;
  params?: { [s: string]: any };
  requestId?: string;
  playground?: boolean;
  cacheTime?: string;
}

export type DataQueryError = {
  error: {
    cancelled?: boolean;
    data: {
      error?: {
        message: string;
        notFound?: string[];
      };
    };
    status: number;
  };
};

export type QueriesData = {
  items: DataQueryRequestItem[];
  target: QueryTarget;
}[];

export type ResponseMetadata = { labels: string[]; target: QueryTarget };

export interface DataQueryAlias {
  alias: string;
  id: number;
  aggregate?: string;
  granularity?: string;
}

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export type DataQueryRequestItem = {
  expression?: string;
  start?: string | number;
  end?: string | number;
  limit?: number;
  granularity?: string;
  aggregates?: string[];
  id?: number;
};

export type Aggregates = Pick<DataQueryRequest, 'aggregates'>;
export type Granularity = Pick<DataQueryRequest, 'granularity'>;

export interface DataQueryRequest {
  items: DataQueryRequestItem[];
  start?: string | number;
  end?: string | number;
  limit?: number;
  aggregates?: string[];
  granularity?: string;
}

export interface Annotation {
  datasource: string;
  enable: boolean;
  hide: boolean;
  iconColor: string;
  limit: number;
  name: string;
  query: string;
  filter: string;
  error: string;
  type: string;
  tags: string[];
}

export interface AnnotationQueryOptions {
  range: GrafanaTimeRange;
  rangeRaw: GrafanaTimeRange;
  annotation: Annotation;
  dashboard: number;
}

export interface AnnotationResponse {
  annotation: Annotation;
  title: string;
  time: number;
  timeEnd?: number;
  text: string;
  tags?: string[];
  isRegion?: boolean;
}

export interface AnnotationSearchQuery {
  description: string;
  type: string;
  subtype: string;
  minStartTime: number;
  maxStartTime: number;
  minEndTime: number;
  maxEndTime: number;
  minCreatedTime: number;
  maxCreatedTime: number;
  minLastUpdatedTime: number;
  maxLastUpdatedTime: number;
  // format is {"k1": "v1", "k2": "v2"}
  metadata: string;
  assetIds: number[];
  assetSubtrees: number[];
  sort: 'startTime' | 'endTime' | 'createdTime' | 'lastUpdatedTime';
  dir: 'asc' | 'desc';
  limit: number;
  offset: 0;
}

export interface Event {
  id: number;
  startTime: number;
  endTime: number;
  description: string;
  type: string;
  subtype: string;
  assetIds: number[];
  source: string;
  sourceId: string;
}

export type Events = Items<Event>;

export type DataEvents = DataResponse<Events>;

export type AnnotationQueryRequestResponse = DataResponse<DataEvents>;

export interface DataResponse<T> {
  data: T;
}

export type Items<T = object> = {
  items: T[];
};

export type CursorResponse<T> = DataResponse<Items<T> & { nextCursor?: string }>;

export type Response<T = object> = DataResponse<{
  items: T[];
}>;

export type TimeseriesFilterQuery = {
  filter?: {
    description?: string;
    assetSubtreeIds?: IdEither[];
    assetIds?: string[];
    [s: string]: any; // so we auto-support next features
  };
  cursor?: string;
} & Limit;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface VariableQueryData {
  query: string;
  error?: string;
}

export interface VariableQueryProps {
  query: string;
  onChange: (query: VariableQueryData) => void;
  datasource: any;
  templateSrv: any;
}

export interface CogniteDataSourceSettings extends DataSourceSettings {
  jsonData: {
    authType: string;
    defaultRegion: string;
    cogniteProject: string;
  };
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter: Filter;
}

export interface QueryRequestError {
  refId: string;
  error: string;
}

export interface QueryDatapointsLimitWarning {
  refId: string;
  warning: string;
}
