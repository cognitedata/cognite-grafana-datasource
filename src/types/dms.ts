// DMS API types
export interface DMSSpace {
  space: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSView {
  space: string;
  externalId: string;
  version: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSInstanceId {
  space: string;
  externalId: string;
}

export interface DMSFilter {
  and?: DMSFilter[];
  or?: DMSFilter[];
  not?: DMSFilter;
  equals?: {
    property: string[];
    value: any;
  };
  in?: {
    property: string[];
    values: any[];
  };
  range?: {
    property: string[];
    gte?: any;
    gt?: any;
    lte?: any;
    lt?: any;
  };
}

export interface DMSSearchRequest {
  view: {
    type: 'view';
    space: string;
    externalId: string;
    version: string;
  };
  query?: string;
  filter?: DMSFilter;
  limit?: number;
  properties?: string[];
}

export interface DMSSearchResponse {
  items: DMSInstance[];
}

export interface DMSInstance {
  instanceType: 'node' | 'edge';
  space: string;
  externalId: string;
  version: number;
  lastUpdatedTime: number;
  createdTime: number;
  deletedTime?: number;
  properties?: {
    [space: string]: {
      [view: string]: {
        [property: string]: any;
      };
    };
  };
}
