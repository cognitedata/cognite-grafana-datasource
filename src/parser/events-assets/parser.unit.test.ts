import { parseQuery, formatQueryParse, parse } from './index';
import { QueryParserTypes } from '../types';

describe('Query parser', () => {
  describe('Query parsing', () => {
    it('should parse string', () => {
      const assetsInput = `assets{rootIds={id=123}, description="description"}`;
      const eventsInput = `events{description=~'some.*', assetIds=[123]}`;
      const { type: assetType, query: assetQuery } = parseQuery(assetsInput);
      const { type: eventsType, query: eventsQuery } = parseQuery(eventsInput);

      expect(assetType).toEqual(QueryParserTypes.assets);
      expect(assetQuery.length).toEqual(2);

      expect(eventsType).toEqual(QueryParserTypes.events);
      expect(eventsQuery.length).toEqual(2);
    });
    it('should throw error in case of empty query', () => {
      const input = ``;

      expect(() => parseQuery(input)).toThrowErrorMatchingSnapshot();
    });
    it('should throw error if wrong type provided', () => {
      const input = `asset{name='name'}`;

      expect(() => parseQuery(input)).toThrowErrorMatchingSnapshot();
    });
    it('should throw error if wrong filter provided', () => {
      const input = `assets{name~='name'}`;

      expect(() => parseQuery(input)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('Formatting', () => {
    it('should return filters', () => {
      const parsed = {
        type: QueryParserTypes.events,
        query: [
          {
            assetIds: [123],
            metadata: {
              'key=~': {
                key: 'key1',
                filter: '=~',
                value: 'test.*',
              },
              key2: 'test',
            },
          },
        ],
      };

      const {
        type,
        params: { metadata, assetIds },
        filters: [filter],
      } = formatQueryParse(parsed);

      expect(type).toEqual(QueryParserTypes.events);
      expect(metadata).toEqual({ key2: 'test' });
      expect(assetIds).toEqual([123]);
      expect(filter).toEqual({ path: 'metadata.key1', filter: '=~', value: 'test.*' });
    });
  });

  describe('Parser result', () => {
    it('should parse empty request', () => {
      const assets = `assets{}`;
      const events = `events{}`;

      const { type: assetsType, params: assetsParams, filters: assetFilters } = parse(assets);
      const { type: eventsType, params: eventsParams, filters: eventsFilters } = parse(events);

      expect(assetsType).toEqual(QueryParserTypes.assets);
      expect(assetsParams).toEqual({});
      expect(assetFilters).toEqual([]);

      expect(eventsType).toEqual(QueryParserTypes.events);
      expect(eventsParams).toEqual({});
      expect(eventsFilters).toEqual([]);
    });

    it('should parse filters and params', () => {
      const query = `assets{rootIds=[{id=123}], description!='non-description'}`;
      const { params, filters } = parse(query);

      expect(params).toEqual({ rootIds: [{ id: 123 }] });
      expect(filters).toEqual([
        {
          path: 'description',
          filter: '!=',
          value: 'non-description',
        },
      ]);
    });

    it('should parse inherited filters', () => {
      const query = `assets{metadata={key1='value1', key2!~'value2'}}`;
      const {
        params,
        filters: [filter],
      } = parse(query);

      expect(params).toEqual({ metadata: { key1: 'value1' } });
      expect(filter).toEqual({
        path: 'metadata.key2',
        filter: '!~',
        value: 'value2',
      });
    });

    it('should parse strict inequality filters with numbers', () => {
      const query = `assets{id!=1}`;
      const {
        filters: [filter],
      } = parse(query);

      expect(filter).toEqual({
        path: 'id',
        filter: '!=',
        value: 1,
      });
    });

    it('should produce right number of filters', () => {
      const query = `assets{metadata={key1='value1', key2!~'value2', key2!='value', key2=~'value'}}`;

      const { filters } = parse(query);
      expect(filters.length).toEqual(3);
    });

    it('should remove empty params', () => {
      const query = `assets{name='name', metadata={key2!~'value2'}}`;
      const {
        params: { name, metadata },
        filters,
      } = parse(query);
      expect(name).toEqual('name');
      expect(filters.length).toEqual(1);
      expect(metadata).toBeFalsy();
    });

    it('should be space agnostic', () => {
      const query = ` assets{ name = 'name' , metadata = { key2 !~ 'value2'}  }  `;
      const queryWithoutSpaces = query.replace(' ', '');
      const parsedWithSpaces = parse(query);
      const parsedWithoutSpaces = parse(queryWithoutSpaces);

      expect(parsedWithSpaces).toEqual(parsedWithoutSpaces);
    });
    it('should parse escape characters in single quoted string', () => {
      const { raw: r } = String;
      const querySBS = r`assets{metadata={key1!~'\value2 \d \. - \w+', key2!~"\value2 \d \. - \w+"}}`;
      const queryDBS = r`assets{metadata={key1=~'\\value2', key2=~"\\value2"}}`;
      const escapeQuote = r`assets{metadata={key1=~'\\value\'2', key2=~"\\value\"2"}}`;
      const escapeQuoteBS = r`assets{metadata={key1=~'\\value\\'2', key2=~"\\value\\"2"}}`;
      const {
        filters: [filterSBS1, filterSBS2],
      } = parse(querySBS);
      const {
        filters: [filterDBS1, filterDBS2],
      } = parse(queryDBS);
      const {
        filters: [filterEscapeQuote1, filterEscapeQuote2],
      } = parse(escapeQuote);
      const {
        filters: [filterEscapeQuoteBS1, filterEscapeQuoteBS2],
      } = parse(escapeQuoteBS);

      expect(filterSBS1).toEqual({
        path: 'metadata.key1',
        filter: '!~',
        value: r`\value2 \d \. - \w+`,
      });
      expect(filterSBS2).toEqual({
        path: 'metadata.key2',
        filter: '!~',
        value: r`\value2 \d \. - \w+`,
      });
      expect(filterDBS1).toEqual({
        path: 'metadata.key1',
        filter: '=~',
        value: r`\\value2`,
      });
      expect(filterDBS2).toEqual({
        path: 'metadata.key2',
        filter: '=~',
        value: r`\\value2`,
      });
      expect(filterEscapeQuote1).toEqual({
        path: 'metadata.key1',
        filter: '=~',
        value: r`\\value\'2`,
      });
      expect(filterEscapeQuote2).toEqual({
        path: 'metadata.key2',
        filter: '=~',
        value: r`\\value\"2`,
      });
      expect(filterEscapeQuoteBS1).toEqual({
        path: 'metadata.key1',
        filter: '=~',
        value: r`\\value\\'2`,
      });
      expect(filterEscapeQuoteBS2).toEqual({
        path: 'metadata.key2',
        filter: '=~',
        value: r`\\value\\"2`,
      });
    });
    it('should parse special cases', () => {
      const { raw: r } = String;
      const query = r`assets{name="\n some", type=~"\n some"}`;
      const {
        params: { name },
        filters: [filter],
      } = parse(query);
      expect(filter).toEqual({
        path: 'type',
        filter: '=~',
        value: r`\n some`,
      });
      expect(name).toEqual('\n some');
    });
  });
});
