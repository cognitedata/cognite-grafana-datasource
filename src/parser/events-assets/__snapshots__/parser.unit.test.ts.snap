// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query parser Query parsing should throw error if wrong filter provided 1`] = `
"Parser: Syntax error:
assets{name~='name'}
           ^"
`;

exports[`Query parser Query parsing should throw error if wrong type provided 1`] = `
"Parser: Syntax error:
asset{name='name'}
     ^"
`;

exports[`Query parser Query parsing should throw error in case of empty query 1`] = `
"Parser: Unexpected end of input:

^"
`;
